src/qat/purr/compiler/instructions.py

```diff
@@ -741,3 +752,137 @@ def calculate_duration(instruction, return_samples: bool = True):
         calc_sample = block_number * block_time
 
     return calc_sample
+
+
+class InstructionBlock:
+    def _validate_targets(self, targets, valid_types):
+        if targets is None:
+            targets = []
+        elif not isinstance(targets, List):
+            targets = [targets]
+
+        invalid_targets = [val for val in targets if not isinstance(val, valid_types)]
+        if any(invalid_targets):
+            invalid_targets_str = ",".join([str(val) for val in invalid_targets])
+            raise ValueError(f"Invalid targets for {type(self)}: {invalid_targets_str}")
+
+        return targets
+
+
+class QuantumInstructionBlock(InstructionBlock):
+    quantum_targets: List[QuantumComponent]
+    instructions: List[Instruction]
+
+
+class MeasureBlock(QuantumInstructionBlock):
+    def __init__(
+        self,
+        targets: Union[Qubit, List[Qubit]],
+        mode: AcquireMode,
+        output_variables: Union[str, List[str]] = None,
+        entangled_qubits: List[Qubit] = None,
+        existing_names: Set[str] = None,
+    ):
+        self._target_dict = {}
+        self._entangled_qubits = set()
+        self._existing_names = existing_names
+        self.add_measurements(targets, mode, output_variables, entangled_qubits)
+
+    def add_measurements(
+        self,
+        targets: Union[Qubit, List[Qubit]],
+        mode: AcquireMode,
+        output_variables: Union[str, List[str]] = None,
+        entangled_qubits: List[Qubit] = None,
+        existing_names: Set[str] = None,
+    ):
+        targets = self._validate_targets(targets, (Qubit))
+        if len((duplicates := [t for t in targets if t in self.quantum_targets])) > 0:
+            raise ValueError(
+                "Target can only be measured once in a 'MeasureBlock'. "
+                f"Duplicates: {duplicates}"
+            )
+        if not isinstance(output_variables, list):
+            output_variables = [] if output_variables is None else [output_variables]
+        if (num_out_vars := len(output_variables)) == 0:
+            output_variables = [None] * len(targets)
+        elif num_out_vars != len(targets):
+            raise ValueError(
+                f"Unsupported number of `output_variables`: {num_out_vars}, "
+                f"must be `None` or match numer of targets: {len(targets)}."
+            )
+        for target, output_variable in zip(targets, output_variables):
+            meas, acq = self._generate_measure_acquire(
+                target, mode, output_variable, existing_names
+            )
+            self._target_dict[target.full_id()] = {
+                "target": target,
+                "mode": mode,
+                "output_variable": output_variable,
+                "measure": meas,
+                "acquire": acq,
+            }
+        self._entangled_qubits.update(
+            self._validate_targets(entangled_qubits, (Qubit))
+            if entangled_qubits is not None
+            else targets
```

Comment:
Is there a requirement for targets that share a measurement block to be treated as entangled?

Reply:
Not per se. However the normal operation is that `entangled_qubits` for a `qubit` includes the qubit itself. This is a safeguard to ensure that at least all measured qubits are synced before and after measurements.